src/dmabuf.rs

```diff
@@ -0,0 +1,240 @@
+use crate::globals::GlobalData;
+use std::{mem, sync::Mutex};
+use wayland_client::{
+    globals::GlobalList,
+    protocol::{wl_buffer, wl_surface},
+    Connection, Dispatch, Proxy, QueueHandle, WEnum,
+};
+use wayland_protocols::wp::linux_dmabuf::zv1::client::{
+    zwp_linux_buffer_params_v1,
+    zwp_linux_dmabuf_feedback_v1::{self, TrancheFlags},
+    zwp_linux_dmabuf_v1,
+};
+
+pub struct DmabufState {
+    zwp_linux_dmabuf: Option<zwp_linux_dmabuf_v1::ZwpLinuxDmabufV1>,
+    formats: Vec<u32>,
+    modifiers: Vec<(u32, u64)>,
+}
+
+struct DmabufFeedbackTranche {
+    device: Vec<u8>, // dev_t
+    flags: WEnum<TrancheFlags>,
+    formats: Vec<u16>, // indices
+}
+
+#[derive(Default)]
+pub struct DmabufFeedback {
+    format_table: Vec<(u32, u32)>, // TODO: read from fd? Format/modifer pair
+    main_device: Vec<u8>,          // dev_t
+    tranches: Vec<DmabufFeedbackTranche>,
+}
+
+#[derive(Default)]
+pub struct DmabufFeedbackData {
+    pending: Mutex<DmabufFeedback>,
+}
+
+pub struct DmaBufferData;
+
+impl DmabufState {
+    pub fn new<State>(globals: &GlobalList, qh: &QueueHandle<State>) -> Self
+    where
+        State: Dispatch<zwp_linux_dmabuf_v1::ZwpLinuxDmabufV1, GlobalData> + 'static,
+    {
+        // XXX minimum version? how to handle older?
+        let zwp_linux_dmabuf = globals.bind(qh, 1..=4, GlobalData).ok();
```

Comment:
This also seems like a broader stability issue with sctk. If we only implement version 3 here, is it a breaking change to move to version 4? Since users of sctk won't be prepared to deal with dmabuf feedback, and `format` and `modifiers` are no longer sent in version 4. Possibly those could be emulated with default feedback. While for other protocols, calls that used to work may become protocol errors.

It would be annoying to make things like `DmabufState::new` take a version parameter, but it would help with this sort of thing.

Reply:
Yeah the pains of protocols making huge semantic changes in version gaps.

Versions 1-3 and version 4 differ significantly enough that I'd consider those version ranges as entirely different protocols.

Also isn't zwp-linux-dmabuf not very useful without modifiers? (version 3)?

Reply:
Even with small differences between versions, it's hard to deal with since SCTK is too low level to really abstract over any sort of breaking change. And I guess API stability for Wayland protocol versions seems to basically just mean not changing the signatures of events and requests.

> Also isn't zwp-linux-dmabuf not very useful without modifiers? (version 3)?

It's possible to create a buffer without an explicit modifier. I assume if the compositor implements version < 3, the default for a format without specifying a modifier should work? But that isn't guaranteed to work in later versions unless it's listed as one of the valid modifiers?

In particular it notes for the `modifiers` event:

> For legacy support, DRM_FORMAT_MOD_INVALID (that is, modifier_hi == 0x00ffffff and modifier_lo == 0xffffffff) is allowed in this event. It indicates that the server can support the format with an implicit modifier. When a plane has DRM_FORMAT_MOD_INVALID as its modifier, it is as if no explicit modifier is specified. The effective modifier will be derived from the dmabuf.

So version 2 should be equivalent to version 3, but if `modifer` is emitted for every format with `DRM_FORMAT_MOD_INVALID`? In which case supporting version 1-2 isn't particularly hard if version 3 is supported.